plugins/kubernetes-backend/src/service/KubernetesProxy.ts

```diff
@@ -0,0 +1,290 @@
+      return clusters;
+    } catch (e: any) {
+      this.logger.error(
+        `error with encoded cluster header: ${JSON.stringify(e)}`,
+      );
+    }
+    return {};
+  }
+
+  protected async getClusterDetails(
+    clusterSupplier: KubernetesClustersSupplier,
+    krc: KubernetesProxyClusters,
+  ): Promise<ClusterDetails[]> {
+    const clusters = await clusterSupplier.getClusters();
+
+    const clusterNames = Object.keys(krc);
+
+    const clusterDetails = clusters.filter(c => clusterNames.includes(c.name));
+
+    const clusterDetailsAuth = await Promise.all(
+      clusterDetails.map(async c => {
+        const authMetadata: KubernetesRequestAuth = {
+          token: krc[c.name],
+        };
+        const cAuth = await this.configureAuth(c, authMetadata);
+        return cAuth;
+      }),
+    );
+
+    return clusterDetailsAuth;
+  }
+
+  protected getClusterIP(details: ClusterDetails): string {
+    const client = this.getKubeConfig(details);
+    return client.getCurrentCluster()?.server || '';
+  }
+
+  protected async makeRequestToCluster(
+    details: ClusterDetails,
+    req: Request,
+  ): Promise<KubernetesProxyResponse> {
+    const serverIP = this.getClusterIP(details);
+    if (!serverIP) {
+      this.logger.error(
+        `could not find server IP for cluster with name ${details.name}`,
+      );
+      return {
+        code: ERROR_INTERNAL_SERVER,
+        data: null,
+      };
+    }
+
+    const query = decodeURIComponent(req.params.encodedQuery) || '';
+    const uri = `${serverIP}/${query}`;
+
+    const res = await this.sendClusterRequest(
+      details,
+      uri,
+      req.method,
+      req.body,
+    );
+
+    return res;
+  }
+
+  protected async sendClusterRequest(
+    details: ClusterDetails,
+    uri: string,
+    method: string,
+    body?: any,
+  ): Promise<KubernetesProxyResponse> {
+    const bearerToken = details.serviceAccountToken;
+    if (!bearerToken) {
+      this.logger.error('could not find bearer token!');
+      return {
+        code: ERROR_BAD_REQUEST,
+        data: {
+          error: 'Invalid service account token',
+        },
+      };
+    }
+
+    const reqData: any = {
+      method,
+      headers: {
+        'Content-Type': 'application/json',
+        Authorization: `Bearer ${bearerToken}`,
+      },
+    };
+
+    if (!details.skipTLSVerify) {
+      if (details.caData) {
+        const ca = bufferFromFileOrString('', details.caData).toString();
+        reqData.agent = new https.Agent({ ca });
+      } else {
+        this.logger.info('could not find CA certificate!');
+        return {
+          code: ERROR_INTERNAL_SERVER,
+          data: {
+            error: 'Invalid CA certificate configured within Backstage',
+          },
+        };
+      }
+    }
+
+    if (body && Object.keys(body).length > 0) {
+      reqData.body = JSON.stringify(body);
+    }
+
+    try {
+      const req = await fetch(uri, reqData);
+
+      const res = await req.json();
```

Comment:
There's a bit of a JSON bias here -- some core k8s APIs (like logs, port-forward and exec obviously) do not speak JSON so this implementation can't support those, and I'm pretty confident that clusters can contain [API aggregation layer](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/) services which I think are super flexible and not necessarily required to speak JSON.